tools/run_tests/run_xds_tests.py

```diff
@@ -478,57 +478,77 @@ def test_round_robin(gcp, backend_service, instance_group):
 
 
 def test_secondary_locality_gets_no_requests_on_partial_primary_failure(
-        gcp, backend_service, primary_instance_group,
-        secondary_zone_instance_group):
+        gcp, backend_service, primary_instance_group, secondary_instance_group):
     logger.info(
         'Running test_secondary_locality_gets_no_requests_on_partial_primary_failure'
     )
     try:
         patch_backend_instances(
             gcp, backend_service,
-            [primary_instance_group, secondary_zone_instance_group])
+            [primary_instance_group, secondary_instance_group])
         wait_for_healthy_backends(gcp, backend_service, primary_instance_group)
         wait_for_healthy_backends(gcp, backend_service,
-                                  secondary_zone_instance_group)
-        primary_instance_names = get_instance_names(gcp, instance_group)
-        secondary_instance_names = get_instance_names(
-            gcp, secondary_zone_instance_group)
+                                  secondary_instance_group)
+        # Ensure the backend patch has propagated before verifying which IG is
+        # primary. An unconditional wait is necessary since, prior to the patch,
+        # traffic can only go to primary_instance_group regardless of TD's
+        # assignment post-patch.
+        time.sleep(_WAIT_FOR_VALID_CONFIG_SEC)
```

Comment:
Is this really the only way? It seems like this could introduce a flake. What if it takes longer than this for the config to propagate?

Reply:
~~So this check is essentially trying to prove that, when the config propagates, traffic will *not* go to `secondary_instance_group`. Since that definitely holds before propagation and should continue to hold afterwards, there's not really anything we can wait for here...I guess we could exit early in the (uncommon) case where traffic does start going to `secondary_instance_group` and save some of the unconditional 60 second wait. But the case that occurs 99+% of the time, with all traffic continuing to go to the expected primary locality, does as best as I can tell require this unconditional waiting.~~

Hmm. Actually, maybe it would be preferable just to more-or-less wrap this entire test's asserts into a try-except block - so we begin with the assumption that the expected primary zone really is primary and, only if that fails (which will be very rare), fall back to check if the test would have passed if we swapped our primary/secondary expectation. This avoids the silly unconditional wait and keeps things "speedy" for the common case.

I'll try to rework things and update this PR.

Reply:
Thanks for the review and helpful feedback here! Closing this in favor of https://github.com/grpc/grpc/pull/23206 (sorry for the churn)